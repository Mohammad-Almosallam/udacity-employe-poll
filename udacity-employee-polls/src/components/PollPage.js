import {
  InputLeftAddon,
  Input,
  InputGroup,
  Image,
  Button,
  Box,
  Flex,
  Stat,
  StatLabel,
  StatNumber,
  StatHelpText,
  StatGroup,
  Text,
} from "@chakra-ui/react";
import React from "react";
import Nav from "./Nav";
import { withRouter, formatQuestion } from "../utils/helper";
import { connect } from "react-redux";
import { handleSubmitQuestionAnswer } from "../actions/questions";
function PollPage(props) {
  const { name, hasAnswered, avatar, optionOne, optionTwo } = props.question;

  function calcPercentage(numOfVotes) {
    const numOptionOneVotes = optionOne.votes.length;
    const numOptionTwoVotes = optionTwo.votes.length;

    const total = numOptionOneVotes + numOptionTwoVotes;

    return Math.floor((numOfVotes / total) * 100);
  }

  function handleBtnSubmit(e, answerValue) {
    e.preventDefault();
    const { authedUser, dispatch, question } = props;
    console.log(authedUser, question.id, answerValue);
    dispatch(
      handleSubmitQuestionAnswer({
        authedUser: authedUser,
        qid: question.id,
        answer: answerValue,
      })
    );
  }
  return (
    <>
      <Nav />
      <Flex
        alignItems={"center"}
        textAlign={"center"}
        justifyContent={"center"}
      >
        <Flex w="70%" alignItems={"center"} flexDir={"column"}>
          <Box>
            <Image w={"200px"} m={"auto"} src={avatar} />
          </Box>
          <Box>
            <Text fontSize={"4xl"} fontWeight={"thin"}>
              {name} asks
            </Text>
          </Box>
          <Box>
            <Text fontSize={"4xl"} fontWeight={"bold"}>
              Would you rather...
            </Text>
          </Box>
          <Flex w={"100%"} mt={"3rem"} gap={"1rem"}>
            <Flex w={"100%"} flexDir={"column"} gap={".5rem"}>
              <InputGroup size="md">
                <InputLeftAddon children="A" />
                <Input
                  disabled
                  placeholder="mysite"
                  value={optionOne.text}
                  fontSize={"1.2rem"}
                />
              </InputGroup>
              <Button
                isDisabled={hasAnswered}
                onClick={(e) => {
                  handleBtnSubmit(e, "optionOne");
                }}
                colorScheme={
                  optionOne.votes.includes(props.authedUser)
                    ? "green"
                    : "facebook"
                }
                w={"100%"}
              >
                Choose A
              </Button>
            </Flex>
            <Flex w={"100%"} flexDir={"column"} gap={".5rem"}>
              <InputGroup size="md">
                <InputLeftAddon children="B" />
                <Input
                  disabled
                  placeholder="mysite"
                  value={optionTwo.text}
                  fontSize={"1.2rem"}
                />
              </InputGroup>
              <Button
                isDisabled={hasAnswered}
                onClick={(e) => {
                  handleBtnSubmit(e, "optionTwo");
                }}
                colorScheme={
                  optionTwo.votes.includes(props.authedUser)
                    ? "green"
                    : "facebook"
                }
                w={"100%"}
              >
                Choose B
              </Button>
            </Flex>
          </Flex>
          {hasAnswered && (
            <Flex w={"100%"} mt={"3rem"}>
              <StatGroup w={"inherit"}>
                <Stat>
                  <StatLabel fontSize={"1rem"}>Number of votes</StatLabel>
                  <StatNumber fontWeight={"extrabold"} fontSize={"3rem"}>
                    {optionOne.votes.length}
                  </StatNumber>
                  <StatHelpText fontSize={"1.2rem"} pl={"0.7rem"}>
                    {calcPercentage(optionOne.votes.length) + "%"}
                  </StatHelpText>
                </Stat>

                <Stat>
                  <StatLabel fontSize={"1rem"}>Number of votes</StatLabel>
                  <StatNumber fontWeight={"extrabold"} fontSize={"3rem"}>
                    {optionTwo.votes.length}
                  </StatNumber>
                  <StatHelpText fontSize={"1.2rem"} pl={"0.7rem"}>
                    {calcPercentage(optionTwo.votes.length) + "%"}
                  </StatHelpText>
                </Stat>
              </StatGroup>
            </Flex>
          )}
        </Flex>
      </Flex>
    </>
  );
}

const mapStateToProps = ({ authedUser, questions, users }, props) => {
  const { id } = props.router.params;
  const question = questions[id];
  return {
    question: question
      ? formatQuestion(question, users[question.author], authedUser)
      : null,
    authedUser,
  };
};
export default withRouter(connect(mapStateToProps)(PollPage));
