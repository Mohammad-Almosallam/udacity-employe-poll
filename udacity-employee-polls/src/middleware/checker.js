const checker = (store) => (next) => (action) => {};

export default checker;
