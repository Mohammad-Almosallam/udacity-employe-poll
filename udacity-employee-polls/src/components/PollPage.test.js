import React from "react";
import PollPage from "./PollPage";
import reducer from "../reducers";
import middleware from "../middleware";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { createStore } from "redux";

const store = createStore(reducer, middleware);

describe("PollPage", () => {
  it("renders the PollPage component completely and correctly", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <PollPage />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
  it("will have all expected fields", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <PollPage />
        </BrowserRouter>
      </Provider>
    );
    const optionOneInput = component.getByTestId("optionOneInput");
    const optionTwoInput = component.getByTestId("optionTwoInput");
    expect(optionOneInput).toBeInTheDocument();
    expect(optionTwoInput).toBeInTheDocument();

    const submitButtonA = component.getByText("Choose A");
    const submitButtonB = component.getByText("Choose B");
    expect(submitButtonA).toBeInTheDocument();
    expect(submitButtonB).toBeInTheDocument();
  });
});
