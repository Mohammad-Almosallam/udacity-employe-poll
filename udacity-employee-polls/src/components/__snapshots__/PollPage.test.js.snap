// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollPage renders the PollPage component completely and correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-0"
      >
        <div
          class="css-1rgb79v"
        >
          <div
            class="css-1gippar"
          >
            <a
              href="/"
            >
              Home
            </a>
            <a
              href="/leaderboard"
            >
              Leaderboard
            </a>
            <a
              href="/add"
            >
              New
            </a>
          </div>
          <div
            class="css-1d6oiat"
          >
            <div
              class="css-rvp98e"
            >
              <img
                class="chakra-image css-1u7uhqd"
              />
              <p
                class="chakra-text css-0"
              />
            </div>
            <button
              class="chakra-button css-4wn91n"
              data-testid="dark-btn"
              type="button"
              value="dark"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 48v48m0 320v48m147.08-355.08l-33.94 33.94M142.86 369.14l-33.94 33.94M464 256h-48m-320 0H48m355.08 147.08l-33.94-33.94M142.86 142.86l-33.94-33.94"
                  fill="none"
                  stroke-linecap="round"
                  stroke-miterlimit="10"
                  stroke-width="32"
                />
                <circle
                  cx="256"
                  cy="256"
                  fill="none"
                  r="80"
                  stroke-linecap="round"
                  stroke-miterlimit="10"
                  stroke-width="32"
                />
              </svg>
            </button>
            <button
              class="chakra-button css-yo5twi"
              type="button"
            >
              Logout
              <svg
                fill="currentColor"
                font-size="1.3rem"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M304 336v40a40 40 0 01-40 40H104a40 40 0 01-40-40V136a40 40 0 0140-40h152c22.09 0 48 17.91 48 40v40m64 160l80-80-80-80m-192 80h256"
                  fill="none"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="32"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="css-1nmt4at"
      >
        <div
          class="css-p4x7e5"
        >
          <div
            class="css-0"
          >
            <img
              class="chakra-image css-1tipa8n"
            />
          </div>
          <div
            class="css-0"
          >
            <p
              class="chakra-text css-1h4i67k"
            >
               asks
            </p>
          </div>
          <div
            class="css-0"
          >
            <p
              class="chakra-text css-odt1ir"
            >
              Would you rather...
            </p>
          </div>
          <div
            class="css-3y3hql"
          >
            <div
              class="css-kvajf0"
            >
              <div
                class="chakra-input__group css-bx0blc"
                data-group="true"
              >
                <div
                  class="chakra-input__left-addon css-1lhl5nw"
                >
                  A
                </div>
                <input
                  class="chakra-input css-yu7fw3"
                  data-testid="optionOneInput"
                  disabled=""
                  placeholder="mysite"
                  value=""
                />
              </div>
              <button
                class="chakra-button css-uxt1e8"
                type="button"
              >
                Choose A
              </button>
            </div>
            <div
              class="css-kvajf0"
            >
              <div
                class="chakra-input__group css-bx0blc"
                data-group="true"
              >
                <div
                  class="chakra-input__left-addon css-1lhl5nw"
                >
                  B
                </div>
                <input
                  class="chakra-input css-yu7fw3"
                  data-testid="optionTwoInput"
                  disabled=""
                  placeholder="mysite"
                  value=""
                />
              </div>
              <button
                class="chakra-button css-uxt1e8"
                type="button"
              >
                Choose B
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-0"
    >
      <div
        class="css-1rgb79v"
      >
        <div
          class="css-1gippar"
        >
          <a
            href="/"
          >
            Home
          </a>
          <a
            href="/leaderboard"
          >
            Leaderboard
          </a>
          <a
            href="/add"
          >
            New
          </a>
        </div>
        <div
          class="css-1d6oiat"
        >
          <div
            class="css-rvp98e"
          >
            <img
              class="chakra-image css-1u7uhqd"
            />
            <p
              class="chakra-text css-0"
            />
          </div>
          <button
            class="chakra-button css-4wn91n"
            data-testid="dark-btn"
            type="button"
            value="dark"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 48v48m0 320v48m147.08-355.08l-33.94 33.94M142.86 369.14l-33.94 33.94M464 256h-48m-320 0H48m355.08 147.08l-33.94-33.94M142.86 142.86l-33.94-33.94"
                fill="none"
                stroke-linecap="round"
                stroke-miterlimit="10"
                stroke-width="32"
              />
              <circle
                cx="256"
                cy="256"
                fill="none"
                r="80"
                stroke-linecap="round"
                stroke-miterlimit="10"
                stroke-width="32"
              />
            </svg>
          </button>
          <button
            class="chakra-button css-yo5twi"
            type="button"
          >
            Logout
            <svg
              fill="currentColor"
              font-size="1.3rem"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M304 336v40a40 40 0 01-40 40H104a40 40 0 01-40-40V136a40 40 0 0140-40h152c22.09 0 48 17.91 48 40v40m64 160l80-80-80-80m-192 80h256"
                fill="none"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="32"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-1nmt4at"
    >
      <div
        class="css-p4x7e5"
      >
        <div
          class="css-0"
        >
          <img
            class="chakra-image css-1tipa8n"
          />
        </div>
        <div
          class="css-0"
        >
          <p
            class="chakra-text css-1h4i67k"
          >
             asks
          </p>
        </div>
        <div
          class="css-0"
        >
          <p
            class="chakra-text css-odt1ir"
          >
            Would you rather...
          </p>
        </div>
        <div
          class="css-3y3hql"
        >
          <div
            class="css-kvajf0"
          >
            <div
              class="chakra-input__group css-bx0blc"
              data-group="true"
            >
              <div
                class="chakra-input__left-addon css-1lhl5nw"
              >
                A
              </div>
              <input
                class="chakra-input css-yu7fw3"
                data-testid="optionOneInput"
                disabled=""
                placeholder="mysite"
                value=""
              />
            </div>
            <button
              class="chakra-button css-uxt1e8"
              type="button"
            >
              Choose A
            </button>
          </div>
          <div
            class="css-kvajf0"
          >
            <div
              class="chakra-input__group css-bx0blc"
              data-group="true"
            >
              <div
                class="chakra-input__left-addon css-1lhl5nw"
              >
                B
              </div>
              <input
                class="chakra-input css-yu7fw3"
                data-testid="optionTwoInput"
                disabled=""
                placeholder="mysite"
                value=""
              />
            </div>
            <button
              class="chakra-button css-uxt1e8"
              type="button"
            >
              Choose B
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
